platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt

```diff
@@ -589,6 +600,24 @@ class ConfigurationCacheState(
             readGradleEnterprisePluginManager(gradle)
             readBuildCacheConfiguration(gradle)
             readCacheConfigurations(gradle)
+            if (readBoolean()) {
+                readBuildProfile(gradle)
+            }
+        }
+    }
+
+    private
+    fun DefaultWriteContext.writeBuildProfile(gradle: GradleInternal) {
+        val buildProfile = gradle.serviceOf<BuildProfile>()
+        writeString(buildProfile.buildDir.path)
```

Comment:
Still storing buildDid state, since `Gradle` service that used as a source is only available on `Build` scope, whild BuildProfile belongs to `BuildTree` scope

Reply:
> Still storing buildDid state, since `Gradle` service that used as a source is only available on `Build` scope, while BuildProfile belongs to `BuildTree` scope

🤔 It looks like our "push" model for setting the build directory works by accident (included builds set it first, but the root build comes last and sets the proper one). We can probably use `BuildStateRegistry` to get it (but it looks ugly and violates Demeter's law):
```java
buildStateRegistry.getRootBuild().getProjects().getRootProject().getMutableModel().getServices().get(ProjectLayout.class).getBuildDirectory().getAsFile().get();
```